src/Tools/dotnet-stack/Symbolicate.cs

```diff
@@ -0,0 +1,294 @@
+        }
+
+        private static void SetMetadataReaderDictionary(string filePath)
+        {
+            try
+            {
+                static Stream streamProvider(string sp) => new FileStream(sp, FileMode.Open, FileAccess.Read);
+                using Stream stream = new FileStream(filePath, FileMode.Open, FileAccess.Read);
+                if (stream != null)
+                {
+                    MetadataReaderProvider provider = null;
+                    if (filePath.Contains(".dll"))
+                    {
+                        using PEReader peReader = new PEReader(stream);
+                        if (!peReader.TryOpenAssociatedPortablePdb(filePath, streamProvider, out provider, out string pdbPath))
+                        {
+                            return;
+                        }
+                    }
+                    /*else if (filePath.Contains(".pdb"))
+                    {
+                        provider = MetadataReaderProvider.FromPortablePdbStream(stream);
+                    }*/
+                    else
+                    {
+                        return;
+                    }
+                    MetadataReader reader = provider?.GetMetadataReader();
+                    s_metadataReaderDictionary.Add(Path.GetFileNameWithoutExtension(filePath), reader);
+                }
+            }
+            catch
+            {
+                return;
+            }
+        }
+
+        internal sealed class StackTraceInfo
+        {
+            public string Type;
+            public string Method;
+            public string Param;
+            public string Assembly;
+            public string Pdb;
+            public string Token;
+            public string Offset;
+        }
+
+        private static string TrySymbolicateLine(string line)
+        {
+            string ret = line;
+            Match match = s_regex.Match(line);
+            if (!match.Success)
+            {
+                return line;
+            }
+
+            StackTraceInfo stInfo = new StackTraceInfo()
+            {
+                Type = match.Groups["type"].Value,
+                Method = match.Groups["method"].Value,
+                Param = match.Groups["params"].Value,
+                Assembly = match.Groups["filename"].Value,
+                Token = match.Groups["token"].Value,
+                Offset = match.Groups["offset"].Value
+            };
+            stInfo.Pdb = stInfo.Assembly.Contains(".ni.dll") ? stInfo.Assembly.Replace(".ni.dll", ".pdb") : stInfo.Assembly.Replace(".dll", ".pdb");
+
+            return GetLineFromMetadata(TryGetMetadataReader(stInfo.Assembly), ret, stInfo);
+        }
+
+        private static MetadataReader TryGetMetadataReader(string assemblyName)
+        {
+            if (s_metadataReaderDictionary.ContainsKey(assemblyName))
+            {
+                return s_metadataReaderDictionary[assemblyName];
+            }
+            return null;
+        }
```

Comment:
This should have logic like this:

```csharp
if (dict.TryGetValue(assemblyName, out MetadataReader reader))
	return reader;
else
	dict[assemblyName] = /* instantiate new MetadataReader */;
```

This will only instantiate a `MetadataReader` when we ask for it.